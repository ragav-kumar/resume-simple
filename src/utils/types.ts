export const endpoints = {
	/** Represents all site content that hails from the DB. Might as well load this in one go */
	content: "content",
	/** For PDF Generation */
	pdf: "pdf",
	/** For Contact form submissions */
	contact: "contact",
} as const;

export type Endpoints = keyof typeof endpoints;

export const routes = {
	profile: "/",
	competency: "/competencies",
	education: "/education",
	contact: "/contact",
	history: "/experience",
	pdf: "/pdf",
} as const;

/**
 * No parameters for the content request.
 * Only reason it's not pre-loaded is because of the fancy loading spinner
 */
export type ContentRequest = undefined;
export interface ContentResponse {
	profile: string[];
	competencies: Competency[];
	education: Education[];
	experience: Experience[];
}

export interface CompetencyItem {
	label: string|null;
	text: string;
}
export interface Competency {
	name: string;
	items: CompetencyItem[];
}

export interface Education {
	title: string;
	location: string;
	start: string;
	end: string|null;
	dateOverride: string|null;
}
export interface Experience {
	title: string;
	employer: string;
	location: string;
	start: string;
	end: string|null;
	dateOverride: string|null;
}
