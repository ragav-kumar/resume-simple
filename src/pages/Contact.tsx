import React from "react";

/**
 * Contact form page!
 */
export const Contact = (  ) => {

	return (
		<p>TODO</p>
	);
}
