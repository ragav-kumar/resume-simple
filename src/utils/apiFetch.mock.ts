import { ContentResponse, Endpoints } from "./types";
import reciprocity from "../img/reciprocity.png";
import cityautobahn from "../img/cityautobahn.png";
import acc from "../img/accredited-accommodations.png";

export const apiFetch = async <
	Request extends unknown,
	Response extends unknown
	>( endpoint: Endpoints, _data?: Request ):Promise<Response> => {

	switch (endpoint ) {
		case "pdf":
			return Promise.resolve(process.env.REACT_APP_API_URL! + "/resume.pdf" as Response);
		case "contact":
			return Promise.resolve(true as any as Response);
		case "content":
			return Promise.resolve(await fetchMockContent() as Response);
	}
	return Promise.reject("");
}

const fetchMockContent = async ():Promise<ContentResponse> => {
	const fetchResponse = await fetch(process.env.REACT_APP_API_URL! + "/content.json", {
		method: 'GET',
		headers: {
			'Content-Type': 'application/json',
			'Access-Control-Allow-Origin': "*",
		},
	});
	const json = await fetchResponse.json() as ContentResponse;

	json.portfolio.delivered.forEach(( { imgUrl }, index) => {
		switch (imgUrl ) {
			case "reciprocity":
				json.portfolio.delivered[index].imgUrl = reciprocity;
				break;
			case "cityautobahn":
				json.portfolio.delivered[index].imgUrl = cityautobahn;
				break;
			case "acc":
				json.portfolio.delivered[index].imgUrl = acc;
				break;
		}
	})

	return json;
}
