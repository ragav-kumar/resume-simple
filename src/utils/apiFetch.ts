import { endpoints, Endpoints } from "./types";
import * as process from "process";

const clientId = process.env.CLIENT_ID || "";
const apiUrl = process.env.API_URL || "http://localhost/";

interface FetchResponseSuccess<T extends unknown> {
	success: true;
	data: T;
	error: undefined;
}

interface FetchResponseFailure {
	success: false;
	data: undefined;
	error: string;
}

export const apiFetch = async <
	Request extends unknown,
	Response extends unknown
	>( endpoint: Endpoints, data?: Request ):Promise<Response> => {
	try {
		const body = new URLSearchParams({
			client_id: clientId,
			data: JSON.stringify(data),
		}).toString();

		const response = await fetch(apiUrl + endpoints[endpoint], {
			method: 'POST',
			headers: {
				'Content-Type': 'application/x-www-form-urlencoded',
			},
			body: body,
		});

		const json = (await response.json()) as FetchResponseSuccess<Response> | FetchResponseFailure;
		return json.success ? Promise.resolve(json.data) : Promise.reject(json.error);
	} catch ( e ) {
		console.log(e)
		return Promise.reject("Connection Error");
	}
}