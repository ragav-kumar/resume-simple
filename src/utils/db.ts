import { createContext } from 'react';
import { ContentResponse } from "./types";
/**
 * Mockup of DB data. Replace with a link to a mySQL DB later.
 */

/**
 * All data for a given section
 */
export type SectionData = {
	menuLabel: string;
	header   : string;
	path     : string;
};
/**
 * Menu Sections. Defined here cause it's used in multiple places
 */
export const sections:SectionData[] = [
	{
		menuLabel: "Profile",
		header: "Professional Profile",
		path: "/",
	}, {
		menuLabel: "Competencies",
		header: "Professional Competencies",
		path: "/competencies",
	}, {
		menuLabel: "Education",
		header: "Education",
		path: "/education",
	}, {
		menuLabel: "Experience",
		header: "Employment History",
		path: "/experience",
	},
];

export type DbContextType = ContentResponse;

export const blankContext:DbContextType = {
	profile: [],
	competencies: [],
	education: [],
	experience: [],
}
export const DbContext = createContext<DbContextType>(blankContext);