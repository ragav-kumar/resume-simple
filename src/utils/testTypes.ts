import { FieldHelperProps, FieldHookConfig, FieldInputProps, FieldMetaProps } from "formik";

export type FormikUseFieldArgType = string | FieldHookConfig<unknown>;
export type FormikUseFieldReturnType = [FieldInputProps<unknown>, FieldMetaProps<unknown>, FieldHelperProps<unknown>];
