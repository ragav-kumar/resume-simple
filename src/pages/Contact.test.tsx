import React from "react";
import renderer from 'react-test-renderer';
import { act, render, screen, waitFor } from '@testing-library/react';
import { Contact, ContactFormState } from "./Contact";
import userEvent from "@testing-library/user-event";
import * as Api from "../utils/apiFetch.mock";
import { apiFetch } from "../utils/apiFetch.mock";
import { mocked } from "ts-jest/utils";
import clearAllMocks = jest.clearAllMocks;

const labelNames: { [key in keyof ContactFormState]: RegExp } = {
	name: /Name/i,
	phone: /Phone/i,
	message: /Message/i,
	email: /E-mail/i, // Added a hyphen to distinguish the honeypot...
	xQCDEmailCaravan: /Email/i,
}

const okayData: ContactFormState = {
	name: "Joe Bloggs",
	email: "",
	phone: "123 456 7890",
	message: "Hello World",
	xQCDEmailCaravan: "joe@bloggs.com",
} as const;
const validationFailData: ContactFormState = {
	name: "",
	email: "",
	phone: "",
	message: "",
	xQCDEmailCaravan: "",
} as const;
const honeypotData: ContactFormState = {
	name: "Joe Bloggs",
	email: "joe@bloggs.com",
	phone: "123 456 7890",
	message: "Hello World",
	xQCDEmailCaravan: "joe@bloggs.com",
} as const;

const enterDataThenSubmit = ( data: ContactFormState ) => act(() => {
	for ( const dataKey in data ) {
		if ( data.hasOwnProperty(dataKey) ) {
			const key = dataKey as keyof ContactFormState;
			const dataValue = data[key];
			const label = labelNames[key]
			userEvent.paste(screen.getByLabelText(label), dataValue);
		}
	}
	userEvent.click(screen.getByRole('button', { name: /submit/i }));
});

// Pull the snapshot check outside the describe to allow simple setup and teardown
it('should render correctly', () => {
	const tree = renderer
		.create(<Contact/>)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

describe('<Contact /> when fetch is successful', () => {

	/*jest.mock('../utils/apiFetch.mock', () => ({
		apiFetch: jest.fn().mockImplementation(() => Promise.resolve(true)),
	}));*/

	beforeEach(() => {
		// jest.mock('../utils/apiFetch.mock', () => ({ apiFetch: jest.fn() }));
		const mockFetch = jest.spyOn(Api, 'apiFetch');
		render(<Contact/>);
	});
	afterEach(clearAllMocks);

	it('should trigger validation on incomplete input', async () => {
		enterDataThenSubmit(validationFailData);
		// Since we're expecting validation errors, we use waitFor so they'll arrive eventually
		await waitFor(() => {
			expect(screen.getAllByText(/Please enter your name/i)).toHaveLength(1);
			expect(screen.getAllByText(/Please enter a message/i)).toHaveLength(1);
			expect(screen.getAllByText(/Please enter your email address/i)).toHaveLength(1);
		});
	});

	it('should pass in all field values on submit', async () => {
		enterDataThenSubmit(okayData);
		await waitFor(() => {
			expect(mocked(apiFetch)).toBeCalledWith("contact", {
				...okayData,
				email: okayData.xQCDEmailCaravan,
			});
		})
	});

	/*it('should show success message on successful submit', async () => {
		enterDataThenSubmit(okayData);
	});

	it('should not submit if honeypot set', () => {
		enterDataThenSubmit(honeypotData);
	});*/
});

/*describe('<Contact /> when fetch is not successful', () => {
	it('should show failure message on failed submit', () => {
		enterDataThenSubmit(okayData);
	});
});*/
