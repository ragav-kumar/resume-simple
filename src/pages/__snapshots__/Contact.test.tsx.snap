// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<form
  action="#"
  className="sc-eCstlR iBjlqv"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <p
    className="sc-jSgvzq bPKcny"
  >
    Thanks for your interest! Please fill out the form below to get in touch with me. You can also find my contact information in the downloadable PDF of my resume.
  </p>
  <label
    className="sc-gKseQn hXnWn"
  >
    <div
      className="sc-iBPTik fYVppt"
    >
      Name
    </div>
    <input
      className="sc-gsTEea cPMJOM"
      data-testid="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </label>
  <label
    className="sc-gKseQn hXnWn"
  >
    <div
      className="sc-iBPTik fYVppt"
    >
      Phone
    </div>
    <input
      className="sc-gsTEea cPMJOM"
      data-testid="phone"
      name="phone"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </label>
  <label
    className="sc-gKseQn hXnWn"
  >
    <div
      className="sc-iBPTik fYVppt"
    >
      Email
    </div>
    <input
      className="sc-gsTEea cPMJOM"
      data-testid="xQCDEmailCaravan"
      name="xQCDEmailCaravan"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </label>
  <label
    className="sc-gKseQn hXnWn"
  >
    <div
      className="sc-iBPTik fYVppt"
    >
      Message
    </div>
    <textarea
      className="sc-dlfnuX dSKtwr"
      data-testid="message"
      name="message"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </label>
  <label
    className="sc-bdfBQB aHUYv"
  >
    E-mail
    <input
      autoComplete="off"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Your name here"
      type="text"
      value=""
    />
  </label>
  <div
    className="sc-hKgJUU kcWwiC"
  />
  <button
    className="sc-pGacB ciavHz"
    name="submit"
    type="submit"
  >
    Submit
  </button>
</form>
`;
